[RemoteMirrors] Rewrite the MachO reader path.

dcci opened the pull request:
There are several problems fixed by this commit.
-> It removes the dependency on a system header.
-> It works when the pointer size for host and target don't match (i.e. 32-bits/64-bits and viceversa)
-> It reads the file lazily, mapping only the sections that are needed. The old approach mapped the whole __TEXT segment, causing significant slowdowns when trying to integrate this API in the debugger. Now it's almost instant (10 seconds vs milliseconds).
    
<rdar://problem/42306551>

Commit:
[ReflectionContext] Rewrite the Mach reader path.

There are several problems fixed by this commit.
-> It removes the dependency on a system header.
-> It works when the pointer size for host and target don't match
(i.e. 32-bits/64-bits and viceversa)
-> It reads the file lazily, mapping only the sections that are
needed. The old approach mapped the whole __TEXT segment, causing
significant slowdowns when trying to integrate this API in the
debugger. Now it's almost instant (10 seconds vs milliseconds).

<rdar://problem/42306551>

Commit:
[swift-reflection-dump] Use the new API for MachO.

<rdar://problem/42306551>

Commit:
[Reflection] Be less aggressive with unknown objects.

Instead of asserting, just return false.

dcci commented:
@swift-ci please test

mikeash commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -112,83 +103,119 @@ class ReflectionContext
+        {{TypeRefMdSec.first.first, TypeRefMdSec.first.second}, 0},
+        {{ReflStrMdSec.first.first, ReflStrMdSec.first.second}, 0},
         LocalStartAddress,
         RemoteStartAddress};
```
I'm not quite sure that `LocalStartAddress` and `RemoteStartAddress` make sense anymore. My (limited) understanding is that they are used to translate between an address in the local space and an address in the remote space, which assumes that all of the reflection sections have the same delta between the two address spaces. That was true when the entire TEXT segment was being mapped in one shot, but is not necessarily true with these changes.

It looks like the only code that actually uses those addresses is `getRemoteAddrOfTypeRefPointer` in `TypeRefBuilder.cpp`. It only translates addresses within the `TypeReference` sections. Maybe we can get what we need by just storing the remote address of that one section.

dcci replied:
I spoke with Nick and he recommended to just map the whole `__TEXT` as taking in account the delta between sections may be complicated. This is what I did. If this turns out to be a bottleneck we might reconsider it but I really want to make some progress on the debugger integration.

mikeash replied:
Makes sense to me. Is it actually slower to map everything in one chunk compared to mapping several smaller chunks? I'd have expected each mapping operation to be approximately constant time regardless of size, but I could certainly be wrong.

dcci commented:
Whilst this passes the regular test suite, it fails with validation-test (probably uncovering some bugs in the implementation). I'm going to take a look at it next.

Commit:
[Reflection] Fix lifetime issues and out-of-process mapping.

This comes back to the older approach of mapping all __TEXT, as
taking in account the delta between sections seems a little too
complicated for now. Hopefully we'll be back optimizing this
once we're done with debugger integration.

<rdar://problem/42306551>

Commit:
[Reflection] Remove now dead code. NFCI.

<rdar://problem/42306551>

dcci commented:
@swift-ci please test

dcci commented:
@swift-ci please test

dcci commented:
pretty sure this is not my fault:
```
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/Output/swift-build-self-host.py.tmp.dir/swiftpm/Sources/SPMLLBuild/llbuild.swift:15:19: error: no such module 'llbuild'
18:39:47 @_exported import llbuild
18:39:47                   ^
18:39:47 error: terminated(1): /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/bin/swift-build-tool -f /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/Output/swift-build-self-host.py.tmp.dir/swiftpm/.build/debug.yaml main output:
```

@ddunbar is this a known issue?

dcci commented:
@swift-ci please smoke test linux

dcci commented:
@swift-ci please smoke test linux

ddunbar commented:
@aciidb0mb3r can you take a look at the llbuild thing?

aciidb0mb3r commented:
Should be resolved as of yesterday night.

aciidb0mb3r commented:
@swift-ci please test linux

Commit:
[Reflection] Locate __DATA and push into imageRanges.

This brings the current support to parity to the old one.
Thanks to Mike Ash for noticing this.

<rdar://problem/42306551>

dcci commented:
@swift-ci test and merge

dcci commented:
@swift-ci test and merge

dcci commented:
@swift-ci please smoke test and merge

dcci commented:
@swift-ci please smoke test and merge

clackary commented:
@swift-ci smoke test and merge
